api/app.go

```diff
@@ -0,0 +1,295 @@
+)
+
+func InitApp(r *mux.Router) {
+	l4g.Debug("Initializing app api routes")
+
+	sr := r.PathPrefix("/apps").Subrouter()
+
+	sr.Handle("/oauth2/register", ApiUserRequired(registerOAuthApp)).Methods("POST")
+	sr.Handle("/oauth2/allow", ApiUserRequired(allowOAuth)).Methods("GET")
+	sr.Handle("/oauth2/access_token", ApiAppHandler(getAccessToken)).Methods("POST")
+}
+
+func registerOAuthApp(c *Context, w http.ResponseWriter, r *http.Request) {
+	app := model.AppFromJson(r.Body)
+
+	if app == nil {
+		c.SetInvalidParam("registerOAuthApp", "app")
+		return
+	}
+
+	secret := model.NewId()
+
+	app.ClientSecret = secret
+	app.CreatorId = c.Session.UserId
+
+	if result := <-Srv.Store.App().Save(app); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		app = result.Data.(*model.App)
+		app.ClientSecret = secret
+
+		c.LogAudit("client_id=" + app.Id)
+
+		w.Write([]byte(app.ToJson()))
+		return
+	}
+
+}
+
+func allowOAuth(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.LogAudit("attempt")
+
+	w.Header().Set("Content-Type", "application/x-www-form-urlencoded")
+	responseData := map[string]string{}
+
+	responseType := r.URL.Query().Get("response_type")
+	if len(responseType) == 0 {
+		c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Bad response_type")
+		return
+	}
+
+	clientId := r.URL.Query().Get("client_id")
+	if len(clientId) != 26 {
+		c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Bad client_id")
+		return
+	}
+
+	redirectUri := r.URL.Query().Get("redirect_uri")
+	scope := r.URL.Query().Get("scope")
+	state := r.URL.Query().Get("state")
+
+	var app *model.App
+	if result := <-Srv.Store.App().Get(clientId); result.Err != nil {
+		c.Err = model.NewOAuthAppError("allowOAuth", "server_error", "Error accessing the database")
+		return
+	} else {
+		app = result.Data.(*model.App)
+	}
+
+	callback := app.CallbackUrl
+
+	if len(redirectUri) > 0 {
+		registeredUrl, _ := url.Parse(app.CallbackUrl)
+		redirectUrl, _ := url.Parse(redirectUri)
+
+		if redirectUrl.Scheme != "https" {
+			c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Supplied redirect_uri must be https")
+			return
+		}
+
+		if registeredUrl.Host != redirectUrl.Host {
+			c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Supplied redirect_uri host/port did not match registered callback_url")
+			return
+		}
+
+		callback = redirectUri
+	}
+
+	if responseType != model.AUTHCODE_RESPONSE_TYPE {
+		responseData["redirect"] = callback + "?error=unsupported_response_type&state=" + state
+		w.Write([]byte(model.MapToJson(responseData)))
+		return
+	}
+
+	authData := &model.AuthData{UserId: c.Session.UserId, ClientId: clientId, CreateAt: model.GetMillis(), RedirectUri: redirectUri, State: state, Scope: scope}
+	authData.Code = model.HashPassword(fmt.Sprintf("%v:%v:%v:%v", clientId, redirectUri, authData.CreateAt, c.Session.UserId))
+
+	if result := <-Srv.Store.AuthData().Save(authData); result.Err != nil {
```

Comment:
The spec asks for some semi-weird behaviour here where if an error is encountered we return as a redirect with an error in the url query (I'm doing the redirect on the client) but only if the redirect_uri host/port matches the registered callback_url host. If they don't match, then just return a normal http error code. 

It gets kind of weird since if the request is found to invalid before we check if the redirect_uri matches the callback_url it's ambiguous as to whether we return a regular http error or just redirect with the error in the query string